### PR TITLE
refactor Segmentation

### PR DESCRIPTION
**Pull request by NatanBagrov**
refactored unet and stdc arch to support yaml upsample refactored stdc to accept unet backbone
added a registry for segmentation backbones

### CODE REVIEW

File: src/super_gradients/training/models/segmentation_models/common.py [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
+from abc import ABC, abstractmethod
+
+from typing import List
+
 from super_gradients.modules import ConvBNReLU
 
 
+class AbstractSegmentationBackbone(nn.Module, ABC):
+    @abstractmethod
+    def get_backbone_output_number_of_channels(self) -> List[int]:
```
Comment (BloodAxe) [inline src/super_gradients/training/models/segmentation_models/common.py:11]: Do we want to also have a method `get_backbone_output_strides(self) -> List[int]`

Or maybe even better to make it a bit future-proof:

```
@dataclass
class FeatureMapOutputSpec:
   channels: int
   stride: int
   
def get_output_spec() -> Tuple[FeatureMapOutputSpec, ...]
```

Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/common.py:11]: As I said, there's a lot to improve there :)
I think we can do this one later as part of an additional refactoring

File: src/super_gradients/training/models/segmentation_models/common.py [hunk 2]
```diff
@@ -1,7 +1,20 @@
 import torch.nn as nn
+from abc import ABC, abstractmethod
+
+from typing import List
+
 from super_gradients.modules import ConvBNReLU
 
 
+class AbstractSegmentationBackbone(nn.Module, ABC):
```
[the remark this answers is not in the document]
Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/common.py:9]: Thanks, [ba30f03](https://github.com/Deci-AI/super-gradients/pull/704/commits/ba30f035ff34ed7ef58d1a0e004cb7be338c9e81)

File: src/super_gradients/training/models/segmentation_models/common.py (hunk 1 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/common.py:15]: I agree with @BloodAxe , for this backbone to be a general case for segmentation models, the out strides must be exposed. So that decoder as in STDC or PPLiteSeg could assert for a valid backbone or modify the decoder accordingly

File: src/super_gradients/training/models/segmentation_models/stdc/stdc.py [hunk 3]
```diff
@@ -394,11 +225,15 @@ class STDCSegmentationBase(SgModule):
...
-    @resolve_param("backbone", BaseFactory({"STDCBackbone": STDCBackbone}))
+    @resolve_param("backbone", BaseFactory(SEGMENTATION_BACKBONES))
+    @resolve_param("initial_upsample", TransformsFactory())
+    @resolve_param("final_upsample", TransformsFactory())
     def __init__(
         self,
-        backbone: AbstractSTDCBackbone,
+        backbone: AbstractSegmentationBackbone,
         num_classes: int,
+        initial_upsample: Optional[nn.Upsample],
```
Comment (lkdci) [inline src/super_gradients/training/models/segmentation_models/stdc/stdc.py:235]: upsample can be done by many ways other than `nn.Upsample` such as, ConvTranspose, PixelShuffle...
I don't think we should force the user to pass those modules as part of the constructor, since some of them requires in_channels as argument, thus are defined dynamically.

File: src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml [hunk 4]
```diff
@@ -24,30 +24,39 @@ context_module:
...
 decoder_params:
   # skip expansion ratio value, before fusing the skip features from the encoder with the decoder features, a projection
   # convolution is applied upon the encoder features to project the num_channels by skip_expansion.
   skip_expansion: 0.25
   decoder_scale: 0.25   # num_channels width ratio between encoder stages and decoder stages.
   up_block_types: [UpCatBlock, UpCatBlock, UpCatBlock, UpCatBlock]    # See unet_decoder.UpBlockType for options.
   up_block_repeat_list: [ 1, 1, 1, 1]   # num of blocks per decoder stage, the `block` implementation depends on the up-block type.
-  mode: bilinear
-  align_corners: False
+  mode: ${arch_params.upsample_mode}
```
[the remark this answers is not in the document]
Comment (NatanBagrov) [reply src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml:37]: I agree, anyone can change that anytime.
I put the defaults that were already there.

File: src/super_gradients/training/models/segmentation_models/stdc/stdc.py (hunk 3 above)
Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/stdc/stdc.py:235]: Tried to mimic pre and post scales.
I believe that we can support other upsample models, as long as we can determine (statically) the parameters. I'll think how to align the signatures

File: src/super_gradients/training/models/segmentation_models/common.py (hunk 1 above)
Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/common.py]: Done, [519d66c](https://github.com/Deci-AI/super-gradients/pull/704/commits/519d66c527525dedd8546fec3bd26716903d1c87)

File: src/super_gradients/training/models/segmentation_models/stdc/stdc.py (hunk 3 above)
Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/stdc/stdc.py]: I've loosen up the type-hint, [38d5b64](https://github.com/Deci-AI/super-gradients/pull/704/commits/38d5b64d9fa6bd64cd58e5c0f3af079492ab09df).
So basically any `nn.Module` can fit.

If we want to take it to the next level, we can have some sort of `ReshapeFactory` and register `Upsample`, `ConvTranspose`, `PixelShuffle`, etc... WDYT? 
@lkdci 
@BloodAxe ?

Comment (shaydeci) [reply src/super_gradients/training/models/segmentation_models/stdc/stdc.py]: @NatanBagrov I like your suggestion.

Comment (NatanBagrov) [reply src/super_gradients/training/models/segmentation_models/stdc/stdc.py]: Should it be part of this PR?

**Comment by shaydeci** [comment]
@NatanBagrov @BloodAxe This PR has been open for quite a while now. It has comments that haven't been resolved yet and conflicts.
Please fix and re-review.
